Fix 0x1a6

NicholasCouri opened the pull request:
## Description

The theory Navin has (which is currently the only way I could get a local repro in the end-to-end tests) is that there are scenarios in which a data store might be loading other data stores asynchronously or in response to some event/trigger in between summaries. This will cause the assert 0x1a6 to be hit in case it happens before our SummaryAck. 

## Reviewer Guidance

The current behavior is that assert being hit and we interrupt the refresh, catch the exception and continue the processing. The approach I decided to take is to remove the assert and interrupt the refresh as if we did not have the pending state and reload from the snapshot (refreshLatestSummaryFromSnapshot) in case the current reference number is smaller than the reference number from the SummaryAck. 

Based on the stacks from the telemetry, all the asserts have similar stack to the one we got a local repro for. 

Note: When changing the signature from  refreshLatestSummaryFromPending, I was going to return an enum instead of a boolean but I thought it would be overkill as we simply need to decide whether to interrupt the recursion or not. 

 
## Does this introduce a breaking change?

No

> -   TODO  I will probably check in the test that repros the issue

NicholasCouri committed:
duration should be undefined  when Data_interruped == false

NicholasCouri committed:
Address Build failure

NicholasCouri committed:
Merge branch 'main' into fix_0x1a6

NicholasCouri committed:
Initial attempt to address AB 645 - Assert 0x1a6

vladsud commented:
@NicholasCouri, sharing some thoughts:
1. First of all, I think we should protect the system and reload summarizing container when any errors happen on this path - see https://dev.azure.com/fluidframework/internal/_workitems/edit/893
   - Maybe we will fix one class of problems, but the general flow should be not trying to patch system when we find ourselves in weird situation but return back to well-known state. As long as number of such cases is small, it's much better to go with a bit more expensive flow, but the one that does not introduce new states and new handling.
2. I'd not make any conclusions on the stack - it would always be the same, given that code reacts to summackAck flow (i.e. it's not run as part of some other flow).
3. Summarize nodes are supposed to correctly reflect states through all transitions. These asserts suggest we have a bug somewhere (very likely - in the logic of createChild() API). In other words, the fact that we load data store somewhere should not result in bugs - process of loading nodes in the system should propagate properly state from parent nodes correctly, including all the state about pending and previous summaries.
   - Big suspect here - async nature of refreshLatestSummary(). Any time we have async functionality, it means it is exposed to global state being changed underneath it, while async function runs, including causing reentrancy! While problems might be exposed by loading data store, I'm almost certain two acks very close in time will generate same issues, if processing of first ack did not finish when second started. It feels like system asks for locks - all operations changing structure of nodes (possible - refreshLatestSummary, createChild(), maybe more) should take a lock and prevent concurrent updates (and maybe even - some reads). Or at least start with reporting (telemetry) when reentrancy happens.

Overall, I'd love to see better articulation of the processes happening without summary node. I.e. data store loading is a trigger (maybe), it does not say where / how the bug occurs within summary node structure, and thus it's hard to evaluate if the fix is correct.

vladsud commented:
One thing to add - these asserts happen in between summary op and summary ack.
Looking at ContainerRuntime.submitSummary(), we pause all processing through summary process (all the way till after sending summary op). So, we will not be processing any acks during this time.
Any data store loading is likely happening during summary process itself, not after.
We are likely getting an ack for our own summary - asserts happen right before Summarize_end event. I think we screwed up state of summary node during summary process, while data stores were loaded and nodes where created, not reflecting current (in-flight) summary. It's possible that they started to load during summary and finished load after summary upload is over (and op processing is resumed) - if so, that does not change much, just shows that every async process should assume global state can change while async operation yields. But more likely - all loading happens during summary, to generate search blobs

NicholasCouri commented:
Yes, the telemetry data shows there are no multiple acks being processed and the current summarizer seems to be actually the one that is realizing the data stores while we wait for our own ack. 

Having said that, I believe the original assumption in which the assert was created is not valid anymore (if a node did not have a pending summary, it could not have a lastsummary).  If that is the case, aborting the refreshLatestSummaryFromPending would lead us to simply refresh the summary state fetching from the server and ignoring the local pending summary we have. 
That seems to be the right action to take, isn't it ?  Can we sync on Monday during our 1-1 ? 

==================================================================
The scenario Navin mentioned and for which we can repro the assert, indicates the mixinSummaryHandler, realizes every DataStore at the time of the summary (handle call below) which can lead to the assert:
...
async summarize(...args: any[]) {
            const summary = await super.summarize(...args);
            const content = await handler(this);
            if (content !== undefined) {
                this.addBlob(summary, content.path, content.content);
            }
            return summary;
        }
.. where the handler realizes the DataStore to query for the search content: 

const obj: Provider<SearchContent> = (await DataObject.getDataObject(runtime)) as Provider<SearchContent>;
... 
summary.summary.tree[firstName] = blob;


The components that current implement the SearchContent are: 
ScriptorComponent  (implements ProvideSearchContent)
AtMentionsCollection
DiasCollection
LoopCanvasComponent







indicate

---
In reply to: [1186565242](https://github.com/microsoft/FluidFramework/pull/11144#issuecomment-1186565242) [](http://example.com/codeflow?ancestors=1186565242,1186565242)

tyler-cai-microsoft commented:
Do you happen to have a repro of this bug in our tests?

NicholasCouri commented:
> Do you happen to have a repro of this bug in our tests?

I do now - I will add a review to it.

NicholasCouri commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -286,19 +287,31 @@ export class SummarizerNode implements IRootSummarizerNode {
+                // https://dev.azure.com/fluidframework/internal/_workitems/edit/645
+                return false;
+            } else {
+                // This should only happen if parent skipped recursion AND no prior summary existed.
```
Yes, I will try :)

NicholasCouri committed:
Adding initial version of the SummarizerWithSearch to validate the assert
Moving some of the code from the gc directory to the test-utils so the Summarizer can also make use of it

NicholasCouri commented:
I added the second iteration with the end-to-end test that repros it (based on Navin's repro with search blobs) ; 
I talked to Vlad and he thinks there is a way to prevent us from getting into this state where the assert shows up - I will investigate further before proceeding with the PR. 

 I will also try to investigate if we can repro by using the summarizer onDemand instead of the current summarization the test is doing

NicholasCouri committed:
Merge branch 'main' into fix_0x1a6

NicholasCouri committed:
Updating the refreshLatestSummaryFromPending to address the scenario where we realize a DataStore while waiting for the ack from our summary.
Aftrer comparing the flow for when we find the latest summary vs when we do not find, the only difference is the need of re-creating the last summary node based on the Parent's summaryNode.

NicholasCouri committed:
Add back line removed by accident.

NicholasCouri commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -242,19 +242,40 @@ export class SummarizerNode implements IRootSummarizerNode {
+                    eventName: "SummaryNodeWithoutPendingSummary",
+                    proposalHandle,
+                    referenceSequenceNumber,
+                });
```
Still need to check whether this logging works and perhaps add additional info to help understand the scenario.

NicholasCouri committed:
No need to re-create the SummaryNode as it is a copy of the existing one.

andre4i commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -267,7 +285,9 @@ export class SummarizerNode implements IRootSummarizerNode {
-        this._latestSummary = summaryNode;
+        if (summaryNode !== undefined) {
+            this._latestSummary = summaryNode;
+        }
```
Interesting. So with the current code we'd set the latest summary to undefined if there's no pending summary?

NicholasCouri replied:
I think we still need it for the next summary as it will be updated before adding it to the pendingSummary.

NicholasCouri commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -242,19 +242,37 @@ export class SummarizerNode implements IRootSummarizerNode {
+            if (this._latestSummary !== undefined) {
+                this.wipSummaryLogger?.sendTelemetryEvent({
+                    eventName: "SummaryNodeWithoutPendingSummary",
+                    proposalHandle,
```
Yes, great idea, I will try to add wipReferenceSequenceNumber and more  I'm not sure whether the wipSummaryLogger  works.

NicholasCouri replied:
I think the proposedHandle is not user data - I know we log it on LastSummaryProposedHandleMismatch :)

NicholasCouri committed:
Adding a variable to the SummarizerNode to identify whether or not we originally were waiting for a pendingsummary for that particular node. If we realize a new node, the waitingforpendingSummary will be false and we will confirm the hypothesis we have in which a DataStore is realized in between summary submission and we receiving a SummaryAck.

tyler-cai-microsoft commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts:
```diff
@@ -272,7 +272,7 @@ export class SummarizerNodeWithGC extends SummarizerNode implements IRootSummari
         }
 
-        return super.refreshLatestSummaryFromPending(proposalHandle, referenceSequenceNumber);
+        super.refreshLatestSummaryFromPending(proposalHandle, referenceSequenceNumber);
```
good catch, we return void here

tyler-cai-microsoft commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -242,19 +243,39 @@ export class SummarizerNode implements IRootSummarizerNode {
+                // The only check is to make sure the reference number from the lastSummary is smaller
+                // than the one returned by the ack.
+                // https://dev.azure.com/fluidframework/internal/_workitems/edit/645
+                assert(this._latestSummary.referenceSequenceNumber < referenceSequenceNumber,
```
Now that I think of it. When we summarize we go from container runtime to all the individual datastores to all the individual DDS's. When we are looking at the children, we are only looking at the individual DDSs

NicholasCouri committed:
After talking to Vlad, we agreed to simply update theSummarzer Nodes realized after the submit op with the same data from their parents.

tyler-cai-microsoft commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -242,24 +242,34 @@ export class SummarizerNode implements IRootSummarizerNode {
-                referenceSequenceNumber === summaryNode.referenceSequenceNumber,
-                0x1a7 /* Pending summary reference sequence number should be consistent */,
-            );
+            if (summaryNode === undefined) {
+                // This should only happen if parent skipped recursion AND no prior summary existed.
+                assert(
+                    this._latestSummary === undefined,
+                    0x1a6 /* "Not found pending summary, but this node has previously completed a summary" */,
+                );
+                return;
+            } else {
+                assert(
+                    referenceSequenceNumber === summaryNode.referenceSequenceNumber,
+                    0x1a7 /* Pending summary reference sequence number should be consistent */,
+                );
+
+                // Clear earlier pending summaries
+                this.pendingSummaries.delete(proposalHandle);
+            }
```
Looks like `summaryNode === undefined is done` twice this shouldn't be green right?

NicholasCouri committed:
Fix wrong merge resolution

NicholasCouri committed:
Add smaller repro Tyler created that does not rely on MixinSearch

NicholasCouri committed:
Merge branch 'main' into fix_0x1a6

NicholasCouri committed:
md + lock files

NicholasCouri committed:
When adding thependingSummaries to a child node, we need to make sure the referenceSequenceNumber gets updated as well.

NicholasCouri committed:
Add header to TestUtils

tyler-cai-microsoft commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -242,19 +242,24 @@ export class SummarizerNode implements IRootSummarizerNode {
-                0x1a6 /* "Not found pending summary, but this node has previously completed a summary" */,
-            );
-            return;
+                // This should only happen if parent skipped recursion AND no prior summary existed.
+                assert(
+                    this._latestSummary === undefined,
+                    0x1a6 /* "Not found pending summary, but this node has previously completed a summary" */,
+                );
+                return;
```
Nit: extra spacing

tyler-cai-microsoft commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeUtils.ts:
```diff
@@ -94,11 +94,18 @@ export class SummaryNode {
     public set additionalPath(additionalPath: EscapedPath | undefined) {
         this.summary.additionalPath = additionalPath;
     }
+    public get waitingForPendingSummary(): boolean {
+        return this.summary.waitingForPendingSummary ?? false;
+    }
+    public set waitingForPendingSummary(waitingForPendingSummary: boolean) {
+        this.summary.waitingForPendingSummary = waitingForPendingSummary;
+    }
     constructor(private readonly summary: {
         readonly referenceSequenceNumber: number;
         readonly basePath: EscapedPath | undefined;
         readonly localPath: EscapedPath;
         additionalPath?: EscapedPath;
+        waitingForPendingSummary?: boolean;
```
Nit: Does this still need to be here?

NicholasCouri replied:
Good catch - no we do not - I thought I had removed it!

NicholasCouri committed:
small fixes.

NicholasCouri committed:
Revert changes on GC test.

NicholasCouri committed:
Small fix

tyler-cai-microsoft commented on packages/runtime/runtime-utils/src/summarizerNode/summarizerNode.ts:
```diff
@@ -539,8 +546,23 @@ export class SummarizerNode implements IRootSummarizerNode {
             child.wipReferenceSequenceNumber = this.wipReferenceSequenceNumber;
         }
+        // In case we have pending summaries on the parent, let's initialize it on the child.
+        if (child._latestSummary !== undefined) {
```
Now that I think about it, this block should be moved up into the if statement this.isTrackingInProgress()
